src/Compatibility/ApiCompat/Microsoft.DotNet.ApiCompatibility/Rules/CannotChangeGenericConstraints.cs

```diff
@@ -0,0 +1,167 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Runtime;
+using Microsoft.CodeAnalysis;
+
+namespace Microsoft.DotNet.ApiCompatibility.Rules
+{
+    /// <summary>
+    /// This class implements a rule to check that generic constraints on type parameters do not change.
+    ///   TypeParameter names should not change in strict mode, or when specified
+    ///   strict mode: no changes at all
+    ///   sealed types and non-virtual methods: no additions, removals permitted
+    ///   non-sealed types and virtual methods: no changes
+    /// </summary>
+    public class CannotChangeGenericConstraints : IRule
+    {
+        private readonly IRuleSettings _settings;
+
+        public CannotChangeGenericConstraints(IRuleSettings settings, IRuleRegistrationContext context)
+        {
+            _settings = settings;
+            context.RegisterOnMemberSymbolAction(RunOnMemberSymbol);
+            context.RegisterOnTypeSymbolAction(RunOnTypeSymbol);
+        }
+
+        private void RunOnTypeSymbol(ITypeSymbol? left,
+            ITypeSymbol? right,
+            MetadataInformation leftMetadata,
+            MetadataInformation rightMetadata,
+            IList<CompatDifference> differences)
+        {
+            if (left is not INamedTypeSymbol leftType || right is not INamedTypeSymbol rightType)
+            {
+                return;
+            }
+
+            var leftTypeParameters = leftType.TypeParameters;
+            var rightTypeParameters = rightType.TypeParameters;
```

Comment:
Code style nit: I usually prefer to use the concrete type on the left hand side when it isn't clear from the assignment. This makes reviewing this easier in non IDE scenarios, i.e. GitHub.